Multiple issues with custom-field radio/checkboxes - dev/core/-/issues/6039

vingle opened the pull request:
Overview
----------------------------------------
As identified [here](https://lab.civicrm.org/dev/core/-/issues/6039), there's multiple issues with radio/checkboxes:
1. if no column count is set (ie it's null) the radio/checkbox label+inputs aren't wrapped in a div, and so can wrap orphaning a label (so that the wrong select item appears alone next to a label). Not a bug but a UX danger, especially as the default setting is no column. The wrapping also applies in some contexts but in others it creates a very long line.
2. in the custom field previewer no label appears *unless* no columns are set.
3. also the custom field previewer isn't giving an accurate preview - it shows the older table/td column/radio layout which isn't what appears in contribution pages
4. the contact layout screen preview of these elements can lead to clipping and overlapping.
5. Front-end the radio & checkbox groups aren't have orphaned labels / lack a11y roles/arialabels.

My answer here has been to:
 - change two QuickForm files to wrap all radio/checkboxes in `crm-option-label-pair` regardless if they are set to be in columns or not.
 - table cells containing column-free radio/checkboxes get a div with `.crm-multiple-checkbox-radio-options` class which applies flexbox wrap and gap (and prevents clipping).
 - update custom field preview.php to use the same label display setting used elsewhere.
 - change the preview.php output foreach loop to match the loop elsewhere.
 - remove various elements which ended up getting duplicated as a result of this
 - a few small css tweaks on top.

Before
----------------------------------------

**Contact Layout Editor - no wrapping, and clipping**
<img width="1103" height="269" alt="image" src="https://github.com/user-attachments/assets/d9e55349-7ab6-462d-96e5-e3d38fb4922e" />

**Custom Field Group Preview - four variations - orphaned input/labels, no label on items with columns set**

<img width="605" height="337" alt="image" src="https://github.com/user-attachments/assets/cf184fb5-8740-4016-90c9-94ad642f7aa4" />

**Contribution page + Profile - four variations (works as expected)**

<img width="919" height="493" alt="image" src="https://github.com/user-attachments/assets/6d7b81a1-545f-4fcd-9fa3-b7962307fae1" />

**New Contribution + Profile - several issues**

<img width="1128" height="552" alt="image" src="https://github.com/user-attachments/assets/166f2591-e1a2-4699-89c7-55ad60daf3e5" />

**Contribution page WAVE a11y - 24 alerts**

<img width="663" height="377" alt="image" src="https://github.com/user-attachments/assets/e63006d9-4aff-4fb1-9777-5f27839b2fde" />

After
----------------------------------------

**Contact Layout Editor - four variations**
(Null and 3 column for each of radio & checkbox)

<img width="623" height="536" alt="image" src="https://github.com/user-attachments/assets/2aa153e9-6dfb-4efa-82c7-08e7e225fcb5" />

**Custom Field Group Preview - four variations**
(Null and 3 column for each of radio & checkbox)

<img width="637" height="615" alt="image" src="https://github.com/user-attachments/assets/b57fd5f2-68b4-4c7b-b48b-3626b4359583" />

**Contribution page + Profile - four variations**

<img width="851" height="535" alt="image" src="https://github.com/user-attachments/assets/7e5f0222-586e-4167-8d9b-add4e77e6b23" />

**New Contribution + Profile / null and 2 column** 

<img width="1061" height="305" alt="image" src="https://github.com/user-attachments/assets/f86e0b19-b412-44c6-9b65-5a70d3bbed8a" />

**Contribution page WAVE a11y - 20 alerts**

<img width="618" height="528" alt="image" src="https://github.com/user-attachments/assets/ba0e8824-c145-4c25-8fe5-0dbb60ac6abd" />

**FormBuilder**

<img width="960" height="298" alt="image" src="https://github.com/user-attachments/assets/680c9920-635f-4cf3-95ed-d537a7ff3269" />

Same screens in RiverLea/Minetta:

<img width="756" height="526" alt="image" src="https://github.com/user-attachments/assets/5085af24-ef7e-4afd-97b7-3250b8b4e4dc" />
<img width="813" height="482" alt="image" src="https://github.com/user-attachments/assets/0c9e50f6-46ec-4cf3-b12d-e84cbfa271ac" />

Technical Details
----------------------------------------
For testing I've been looking at:
 - custom fields with checkbox and radio - one with columns set, one without - then preview
 - these fields in a profile on a contribution page - form and confirmation screen
 - these fields in a profile on 'new contribution' layout
 - this field group on the contact layout 
 - these fields in SK/FB (they're re-rendered and all of the above is ignored, so not worth testing)

Frustratingly these all need to be created to test!

Comments
----------------------------------------
Some of the changed lines here are just me changing the indent!

Of the rest, a lot is copy/paste. e.g. https://github.com/civicrm/civicrm-core/compare/master...vingle:civicrm-core:checkbox-issues2?expand=1#diff-20a513cb87e770eddfe52a14917ea6ead2b67550ab5ae480cb2a3eee2a395378R42-R52 where I replaced the table/td layout in preview is taken from `uf/form/fields.tpl` with just a changes to `$formElement`.

Stepping beyond CSS into smarty/quickform is outside my normal comfort zone - I've no idea about wider consequences over what to me seem relatively simple improvements. So I'd appreciate someone reading my changes as well as functionally testing, because this all works in the layouts I tested, and these elements are used in lots of places and probably have extensions/custom JS talking to them.

Final note - two things I didn't manage to fix here - get the spanner icon to appear in the custom fields preview when a column is set, it only appears when one isn't. Also there's a `nbsp;` appearing between the items in Contact Layout (CustomFields.tpl) - and I can't find it anywhere! I deleted a few but it's still appearing and messes with the layout a bit. But these are relatively v small compared to the other fixes in this PR.

vingle committed:
multiple issues with radio/checkboxes - dev/core/-/issues/6039

seamuslee001 commented on templates/CRM/Custom/Form/Preview.tpl:
```diff
@@ -32,49 +32,49 @@
-                 </tr>
-                </table>
-         </td>
+          <td class="label">{if $element.html_type === "Radio"}<span id="{$element_name}_group">{$element.label}</span>{else}{$form.$element_name.label}{/if}{if !empty($element.help_post)}{help id=$element.id file="CRM/Custom/Form/CustomField.hlp" title=$form.$element_name.label}{/if}
```
ok @vingle I manually applied your code on the preview page and using the sample data and setting the Most important issue custom field to be required you will notice that in the attached screenshot when on the preview page there is no red asterisk to indicate it is a required field

<img width="1039" height="465" alt="Screenshot 2025-08-28 103357" src="https://github.com/user-attachments/assets/8a2bf749-5d46-4fce-83a6-f8f9b6d788f2" />

Now changing the label key to be `{$form.$element_name.label}` brings back the red asterisk

<img width="1003" height="401" alt="Screenshot 2025-08-28 103726" src="https://github.com/user-attachments/assets/7819d015-672a-433a-8388-336f92646197" />

However this brings back the Orphaned form label key in wave because you have a <label> key without it being associated with any form element

<img width="1261" height="772" alt="Screenshot 2025-08-28 103801" src="https://github.com/user-attachments/assets/5fdbb39e-9090-4704-a4a5-10b83970784c" />

However having the label key being the following when we are in checkbox or radio seems to produce the red asterisk without the wave issue. 

```
{$form.$element_name.label|regex_replace:"/\<(\/|)label\>/":""}
```

I suspect the $formElement.label was a copy and paste error from mine and I think the preview page may not have been fixed in the pr #33090

seamuslee001 commented on templates/CRM/UF/Form/Fields.tpl:
```diff
@@ -24,20 +24,16 @@
+      <div class="crm-section editrow_{$profileFieldName}-section form-item" id="editrow-{$rowIdentifier}" {if $field.html_type eq 'Radio' or $field.html_type eq 'CheckBox'}role="radiogroup" aria-labelledby="{$profileFieldName}_group"{/if}>
+        <div class="label option-label" {if $field.html_type eq 'Radio' or $field.html_type eq 'CheckBox'}id="{$profileFieldName}_group"{/if}>{$formElement.label}</div>
         <div class="content">
-          <div class="crm-multiple-checkbox-radio-options crm-options-per-line" style="--crm-opts-per-line:{$field.options_per_line};">
```
@vingle does this mean we are no longer respecting the options per line setting on a field or is it being handled elsewhere?

vingle replied:
It's being handled elsewhere. Where? you ask. {blankface}. But this line was appearing twice before this PR so it's an old bug that does unpredictable things with the css.

<img width="910" height="309" alt="image" src="https://github.com/user-attachments/assets/59bcc952-f7c8-46f0-8646-f5b3420c32d0" />

vingle committed:
Fixes asterisk on preview & a11y fail - orphaned labels across checkbox/radio groups

vingle commented:
Having got this far with debugging I've realised you could simplify the three layouts (CustomField.tpl, Preview.tpl, UF/form/Fields.tpl) a bit more. At the moment the code kinda goes:

- Are there options per line set?
    - is it a radio? Do this
    - is it a checkbox? Do this?
- So there's no options per line set, then…
    - is it a radio? Do this
    - is it a checkbox? Do this?
    - ELSE…
 
Instead of:

- Is it a radio?
- Is it a checkbox?
- ELSE…

And then let the CSS handle no option lines set. At the same time - one day this will all be SK/FB, so not so useful focusing on…

vingle commented:
@krisztin - yes, thanks this probably is quite a mess to read!

The div pattern I'm trying to implement [at this line](https://github.com/civicrm/civicrm-core/pull/33466/files#diff-38155304ecb393e20cc7e8cab2aff7a4022b920b1ec621778dca727b4080fe19R26) and [this one](https://github.com/civicrm/civicrm-core/pull/33466/files#diff-38155304ecb393e20cc7e8cab2aff7a4022b920b1ec621778dca727b4080fe19R46) is based on this WAI pattern: https://www.w3.org/WAI/ARIA/apg/patterns/radio/examples/radio/. I had been assuming `{$element_name}_group` would give a unique ID to match label to inputs, removing the need for JS, but perhaps that's not enough.

But this won't handle checked states, and I've not tested tabbing. The only accessibility fix here was the issue of orphaned labels - which because label and inputs are in separate table columns I don't think can be handled with fieldset or a single container div. But very happy for any improvements! I could also remove this part of the PR fix and leave a11y improvements here to a future PR.

Will try to rebuild the demo:

Jenkins test this please.

colemanw commented on templates/CRM/UF/Form/Fields.tpl:
```diff
@@ -24,17 +24,17 @@
-        <div class="label option-label" {if $field.html_type eq 'Radio' or $field.html_type eq 'CheckBox'}id="{$profileFieldName}_group"{/if}>{$formElement.label}</div>
-        <div class="content">
+      <div class="crm-section editrow_{$profileFieldName}-section form-item" id="editrow-{$rowIdentifier}" {if $field.html_type eq 'Radio'}role="radiogroup" aria-labelledby="{$profileFieldName}_group"{/if}>
+        <div class="label option-label" {if $field.html_type eq 'Radio' or $field.html_type eq 'CheckBox'}id="{$profileFieldName}_group">{$formElement.label|regex_replace:"/\<(\/|)label\>/":""}{else}>{$formElement.label}{/if}</div>
```
I'm curious, is this regex intended to get a text-only version of the label? Because this variable is also available: `{$formElement.textLabel}`

vingle replied:
I've just tested it and it works nicely…

vingle replied:
Actually no - that change happened in CustomField.tpl and Preview.tpl - still need to change Fields.tpl - but assume the same principle applies there too

seamuslee001 replied:
@colemanw yeh i was not aware of it and yes the purpose was to get rid of the additional <label> that came from the quickform label element because wave would say that is then an Orphened label

vingle committed:
Removes Regex from Fields.tpl

In favour of textLabel - in line with other two files using this, and as suggested by Coleman.

vingle commented:
Jenkins test this please

vingle committed:
Stack checkbox/radio labels in contact layout

Add styles to stack labels for checkbox/radio in contact layout.
